packages/backend-auth/API.md

```diff
@@ -26,9 +27,9 @@ export type AmazonProviderFactoryProps = Omit<AmazonProviderProps, 'clientId' |
 };
 
 // @public (undocumented)
-export type AmplifyAuthFactoryProps = Omit<AuthProps, 'outputStorageStrategy' | 'loginWith'> & TriggerConfig & {
-    loginWith: AuthLoginWithFactoryProps;
-};
+export type AmplifyAuthFactoryProps = Expand<Omit<AuthProps, 'outputStorageStrategy' | 'loginWith'> & TriggerConfig & {
+    loginWith: Expand<AuthLoginWithFactoryProps>;
+}>;

```

Comment:
From https://github.com/aws-amplify/amplify-backend/pull/606#discussion_r1387325155


Instead of adding more and more type algebra - should we do what data factory did and just define types here and map them to construct types in implementation?

Types without type algebra like this are friendlier to users that don't rely on IDE tool tips to understand what's inside.

Reply:
I don't think data construct has the same issue; using Backend secret as a wrapper for client secrets (Auth has two interfaces, one for cdk users - without BackendSecret support, and one for front end devs - with BackendSecret)

Reply:
I guess we could 'flip' and do the algebra for CDK users

Reply:
trick with that is that construct shouldn't know about its wrapper, and it would if we decided to do algebra in construct

Reply:
> well i guess that we can avoid this by transforming before passing into construct

Yes, that's what I'm thinking, so that types are easy to understand for both populations of users.
I'd say that in this case DX is more important.

Reply:
maybe we should split this out to its own pr, this would impact Hoang's work with https://github.com/aws-amplify/amplify-backend/blob/main/packages/backend-auth/src/translate_auth_props.ts which translate from factory props to construct props and I haven't spent much time with it

Reply:
![image](https://github.com/aws-amplify/amplify-backend/assets/110861985/fcf5e4fe-24e3-4ae1-9d63-42e564b98885)
on the other hand, I can eliminate some of the type algebra ^ because we don't need a "TriggerConfig" type

Reply:
@edwardfoyle @Amplifiyer what do you think?

The https://github.com/aws-amplify/amplify-backend/blob/main/packages/backend-auth/src/translate_auth_props.ts is actually a step forward, it implements some part of the translation we're talking about, so it's not re-work but ~re-use. Btw. `translate_auth_props.ts` is how we'd name a file that has `translate(props: FactoryProps): ConstructProps`.

Reply:
fwiw the type algebra is quite nested; I don't know if non-IDE users should will find this easy either:
![image](https://github.com/aws-amplify/amplify-backend/assets/110861985/75d37ed4-6363-4188-829c-0e6d65a0f279)

Reply:
I sympathize with Kamil's concern, but I also think there's a balance to be struct between type complexity and translation complexity. In this case, I think the `Expand` utility type is less complex than having two separate types and translation between them. The type algebra from the secrets work was significantly trickier hence breaking out into separate types with translation.

Reply:
There's one more thing here:
Sharing types between construct and factory means that construct break is inherently factory's break.

I think we may go for preview with this as is and log release blocker to revisit this. Perhaps users will tell us feedback by then. (if we do this, we should move `Expand` back to auth until we know more).

Reply:
ticket created for this

Reply:
Can we get this in form of github issue ?

Reply:
done https://github.com/aws-amplify/amplify-backend/issues/649